### ISSUE TITLE
No exception when the input format does not match the prescribed type

### ISSUE THREAD

**Issue by asteriosk** [stratosphere/stratosphere#245]
I have played with Scala today. Here is what I have written:

```
case class Department(id: String, name:Int)
...
val departments  = DataSource("departments.txt", RecordDataSourceFormat[Department]("\n", " "))
...
```

The departments.txt file looks like this

```
5 Marketing
2 Enginnering
3 Sales
```

Note the error in my Department class: the id is defined as a String and the department name is an Integer. What I would expect from the "parser" is to tell me that the file does not match what I prescribed. But if you run this program, there will be no exception and the departments input will be just empty. 

I think that the right thing to do is raise an exception saying that "an error was found while parsing the output. Please check your data types." According to @aljoscha , the problem is in the Java side: `RecordInputFormat`.

So, is this a bug or a feature (maybe, ignoring lines that do not match the data source definition)?

**Comment by StephanEwen** [comment]
The reason is that the input format skips invalid line. In the above case, all lines are invalid and skipped.

We can
-  Have a lenient and non-lenient mode (lenient filtering erroneous records, non-lenient throwing errors)
-  Have a way to access to rejected rows...

**Comment by asteriosk** [comment]
Why was this done this way? Did you have a specific use case in mind?

I think we should throw exceptions in the case of bad input: this can really make you spend days debugging if you have a slight problem in your data (or program). 

So, "strict parsing" for me should be the default and we could also provide a constructor that will allow skipping of input lines that are not valid.

What do you think?

**Comment by StephanEwen** [comment]
Sounds reasonable.

I cannot tell you why exactly it is like that, but we should change it definitely.

**Comment by uce** [comment]
Lets also rename lenient to "non-strict" (or to strict and change the code logic).

**Comment by StephanEwen** [comment]
Default behavior is now that parsers fail on invalid lines (710b12ab97ddf8f8f996959afedb1663da3c1252)

**Comment by StephanEwen** [comment]
This issue is fixed in 9dd4635ad69bf107ab9f81c8e7ffd495475471e0

Default behavior is to fail on invalid input. Input format can be set to 'lenient' mode to ignore invalid lines.

@uce I named it 'lenient' for now. I thought that was a rather common term for such behavior. If you think non-strict should be used as a name, feel free to rename it. Leaving the issue open until this is resolved.

**Comment by StephanEwen** [comment]
Resolved
